src/foam/nanos/crunch/Capability.js

```diff
@@ -0,0 +1,200 @@
+    'foam.core.X',
+    'foam.dao.ArraySink',
+    'foam.dao.DAO',
+    'foam.nanos.crunch.Capability',
+    'foam.nanos.crunch.CapabilityCapabilityJunction',
+    'java.util.List',
+    'static foam.mlang.MLang.*',
+  ],
+
+  implements: [
+    'foam.mlang.Expressions',
+  ],
+
+  ids: [
+    'name'
+  ],
+
+  tableColumns: [
+    'name',
+    'description',
+    'version',
+    'enabled',
+    'visible',
+    'expiry',
+    'daoKey'
+  ],
+
+  properties: [
+    
+    {
+      name: 'name',
+      class: 'String',
+      documentation: `name/id of the capability`
+    }, 
+    {
+      name: 'icon',
+      class: 'String',
+      documentation: `path to capability icon`
+    },
+    {
+      name: 'description',
+      class: 'String',
+      documentation: `description of capability`
+    },
+    {
+      name: 'notes',
+      class: 'String'
+    },
+    {
+      name: 'version',
+      class: 'String'
+    },
+    {
+      name: 'enabled',
+      class: 'Boolean',
+      value: true,
+      documentation: `capability is ignored by system when enabled is false.
+      user will lose permissions implied by this capability and upper level capabilities will ignore this prerequisite`
+    },
+    {
+      name: 'visible',
+      class: 'Boolean',
+      documentation: `hide sub-capabilities which aren't top-level and individually selectable. when true, capability is visible to the user`
+    },
+    {
+      name: 'expiry',
+      class: 'DateTime',
+      documentation: `datetime of when capability is no longer valid`
+    },
+    {
+      name: 'of',
+      class: 'Class',
+      documentation: `model used to store information required by this credential`
+    },
+    {
+      name: 'permissionsGranted',
+      class: 'StringArray',
+      documentation: `list of permissions granted by this capability`
+    },
+    {
+      name: 'daoKey',
+      class: 'String',
+      visibility: 'RO'
+    }
+  ],
+
+
+  methods: [
+    {
+      name: 'implies',
+      type: 'Boolean',
+      args: [
+        { name: 'x', type: 'Context' },
+        { name: 'permission', type: 'String' }
+      ],
+      documentation: `checks if a permission or capability string is implied by the current capability`,
+      code: function implies(x, permission) {
+        if( ! this.enabled ) return false;
+
+        if( this.stringImplies(this.name, permission) ) return true;
+
+        this.permissionsGranted.forEach(function(permissionName) {
+          if( permission === permissionName || this.stringImplies(permissionName, permission) ) return true;
+        });
+
+        this.prerequisiteCapabilityJunctionDAO.where(this.EQ(this.CapabilityCapabilityJunction.TARGET_ID, this.id))
+          .select().then(function(sink) {
+            var prerequisites = sink.array
+            prerequisites.forEach(function(prereq) {
+              this.capabilityDAO.find(prereq.sourceId).then(function(cap) {
+                if( cap.implies(x, permission) ) return true;
```

Comment:
This isn't going to work because the inner DAO operations will be async. This method would need to return a promise. Do we even need a JS implementation of this method?

Reply:
Should I remove the js implementation? now that I think about it it doesn't seem like it would be used anywhere